src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs

```diff
@@ -838,14 +737,17 @@ public ProjectState UpdateDocumentsOrder(ImmutableList<DocumentId> documentIds)
                 return this;
             }
 
-            return this.With(
-                projectInfo: this.ProjectInfo.WithVersion(this.Version.GetNewerVersion()),
-                documentIds: documentIds);
+            var reorderedStates = documentIds.Select(id =>
+                DocumentStates.GetState(id) ?? throw new InvalidOperationException($"The document '{id}' does not exist in the project."));
```

Comment:
We don't have a way to keep the dictionary but just reorder the list? Otherwise this will under the covers create a new dictionary that was identical to the old one?